pkg/controller/cronhorizontalpodautoscaler/cronjob.go

```diff
@@ -86,60 +90,156 @@ func (ch *CronJobHPA) Ref() *TargetRef {
 		}
 
-		found := false
-		for _, mapping := range mappings {
-			targetGR = mapping.Resource.GroupResource()
-			scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, ch.TargetRef.RefName)
-			if err == nil {
-				found = true
+		// hpa compatible
+		if ch.TargetRef.RefKind == "HorizontalPodAutoscaler" {
+			msg, err = ch.ScaleHPA()
+			if err != nil {
+				break
+			}
+		} else {
+			msg, err = ch.ScalePlainRef()
+			if err != nil {
 				break
 			}
 		}
+		time.Sleep(updateRetryInterval)
+		times = times + 1
+	}
+
+	return msg, err
+}
+
+func (ch *CronJobHPA) ScaleHPA() (msg string, err error) {
+	var scale *autoscalingapi.Scale
+	var targetGR schema.GroupResource
+
+	ctx := context.Background()
+	hpa := &autoscalingapi.HorizontalPodAutoscaler{}
+	err = ch.client.Get(ctx, types.NamespacedName{Namespace: ch.HPARef.Namespace, Name: ch.TargetRef.RefName}, hpa)
+
+	if err != nil {
+		return "", fmt.Errorf("Failed to get HorizontalPodAutoscaler Ref,because of %v", err)
+	}
+
+	targetRef := hpa.Spec.ScaleTargetRef
+
+	targetGV, err := schema.ParseGroupVersion(targetRef.APIVersion)
+	if err != nil {
+		return "", fmt.Errorf("Failed to get TargetGroup of HPA %s,because of %v", hpa.Name, err)
+	}
+
+	targetGK := schema.GroupKind{
+		Kind:  targetRef.Kind,
+		Group: targetGV.Group,
+	}
+
+	mappings, err := ch.mapper.RESTMappings(targetGK)
+	if err != nil {
+		return "", fmt.Errorf("Failed to create mapping,because of %s", err.Error())
+	}
+
+	found := false
+	for _, mapping := range mappings {
+		targetGR = mapping.Resource.GroupResource()
+		scale, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Get(targetGR, ch.TargetRef.RefName)
+		if err == nil {
+			found = true
+			break
 
-		if !found {
-			return "", fmt.Errorf("Failed to found source target %s", ch.TargetRef.RefName)
 		}
+	}
+
+	if found == false {
+		return "", fmt.Errorf("Failed to found source target %s", ch.TargetRef.RefName)
+	}
+
+	updateHPA := false
 
-		msg = fmt.Sprintf("current replicas:%d, desired replicas:%d", scale.Spec.Replicas, ch.DesiredSize)
-		scale.Spec.Replicas = int32(ch.DesiredSize)
-		_, err = ch.scaler.Scales(ch.TargetRef.RefNamespace).Update(targetGR, scale)
+	if ch.DesiredSize > hpa.Spec.MaxReplicas {
+		hpa.Spec.MaxReplicas = ch.DesiredSize
+		updateHPA = true
+	}
+
+	if ch.DesiredSize < *hpa.Spec.MinReplicas {
+		*hpa.Spec.MinReplicas = ch.DesiredSize
+		updateHPA = true
+	}
+
+	if hpa.Status.CurrentReplicas < ch.DesiredSize {
+		*hpa.Spec.MinReplicas = ch.DesiredSize
+		updateHPA = true
+	}
+
+	if hpa.Status.CurrentReplicas > ch.DesiredSize {
+		// skip change replicas and exit
+		return "", nil
+	}
+
+	if updateHPA {
+		err = ch.client.Update(ctx, hpa)
 		if err != nil {
-			log.Warningf("Failed to scale (namespace: %s;kind: %s;name: %s) to %d,because of %s", ch.TargetRef.RefNamespace, ch.TargetRef.RefKind, ch.TargetRef.RefName, ch.DesiredSize, err.Error())
-		} else {
+			return "", err
+		}
+	}
```

Comment:
```
        if updateHPA {
                err = ch.client.Update(ctx, hpa)
                if err != nil {
                        return "", err
                }
        }

        if hpa.Status.CurrentReplicas > ch.DesiredSize {
                // skip change replicas and exit
                return "", nil
        }
```

I think the *skip* should after the *updateHPA*, or else we won't have a chance to scale down the deployment after scale up the HPA. am i right? 
suppose the initial hpa is (min,desired,max) is (0,3,10), and changed to be (5,5,10) after cronhpa scale up. then scale down to 3 replicas by cronhpa. if we *skip* before *updateHPA*, then we will always be at least 5 replicas.

Reply:
Great, thanks:)